

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>sc-jnmf 1.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home" alt="Documentation Home"> sc-jnmf
          

          
          </a>

          
            
            
          

          

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
            
              <!-- Local TOC -->
              <div class="local-toc"><p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-sc_jnmf">sc_jnmf package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#module-sc_jnmf.sc_jnmf">sc_jnmf.sc_jnmf module</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#module-sc_jnmf">Module contents</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">sc-jnmf</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
        
      <li>sc-jnmf 1.0.0 documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-sc-jnmf-s-documentation">
<h1>Welcome to sc-jnmf’s documentation!<a class="headerlink" href="#welcome-to-sc-jnmf-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<span id="document-sc_jnmf"></span><div class="section" id="sc-jnmf-package">
<h2>sc_jnmf package<a class="headerlink" href="#sc-jnmf-package" title="Permalink to this headline">¶</a></h2>
<div class="section" id="submodules">
<h3>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="module-sc_jnmf.sc_jnmf">
<span id="sc-jnmf-sc-jnmf-module"></span><h3>sc_jnmf.sc_jnmf module<a class="headerlink" href="#module-sc_jnmf.sc_jnmf" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt id="sc_jnmf.sc_jnmf.sc_JNMF">
<em class="property">class </em><code class="sig-prename descclassname">sc_jnmf.sc_jnmf.</code><code class="sig-name descname">sc_JNMF</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">D1</span></em>, <em class="sig-param"><span class="n">D2</span></em>, <em class="sig-param"><span class="n">rank</span></em>, <em class="sig-param"><span class="n">lambda1</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">lambda2</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">lambda3</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">lambda4</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">W1</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">W2</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">H</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">geneset1</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">geneset2</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cluster</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sc_jnmf.sc_jnmf.sc_JNMF" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An analysis tool using Joint-NMF for single cell gene expression profiles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>D1</strong> (<em>pandas DataFrame</em>) – Gene expression matrix of pandas dataframe (row : gene, col : cell).
D1 columns must be same as D2 columns.</p></li>
<li><p><strong>D2</strong> (<em>pandas DataFrame</em>) – Gene expression matrix of pandas dataframe (row : gene, col : cell).
D2 columns must be same as D1 columns.</p></li>
<li><p><strong>rank</strong> (<em>int</em>) – The rank in matrix factorization.</p></li>
<li><p><strong>lambda1</strong> (<em>float</em><em>, </em><em>default 1.0</em>) – The coefficient (parameter) of <a href="#id1"><span class="problematic" id="id2">|</span></a>D2-W2*H|_F^2 in the objective function.</p></li>
<li><p><strong>lambda2</strong> (<em>float</em><em>, </em><em>default 0.0</em>) – The coefficient (parameter) of <a href="#id3"><span class="problematic" id="id4">|W1|</span></a> (l1 or l2 norm) in the objective function.</p></li>
<li><p><strong>lambda3</strong> (<em>float</em><em>, </em><em>default 0.0</em>) – The coefficient (parameter) of <a href="#id5"><span class="problematic" id="id6">|W2|</span></a> (l1 or l2 norm) in the objective function.</p></li>
<li><p><strong>lambda4</strong> (<em>float</em><em>, </em><em>default 1.0</em>) – The coefficient (parameter) of <a href="#id7"><span class="problematic" id="id8">|H|</span></a> (l1 or l2 norm) in the objective function.</p></li>
<li><p><strong>W1</strong> (<em>2d ndarray</em><em> or </em><em>None</em><em>, </em><em>default None</em>) – Initial value of factorized matrix (gene * rank).</p></li>
<li><p><strong>W2</strong> (<em>2d ndarray</em><em> or </em><em>None</em><em>, </em><em>default None</em>) – Initial value of factorized matrix (gene * rank).</p></li>
<li><p><strong>H</strong> (<em>2d ndarray</em><em> or </em><em>None</em><em>, </em><em>default None</em>) – Initial value of factorized matrix (rank * cell).</p></li>
<li><p><strong>geneset1</strong> (<em>None</em>) – The result of ‘gene_selection’ in geneset1.</p></li>
<li><p><strong>geneset2</strong> (<em>None</em>) – The result of ‘gene_selection’ in geneset2.</p></li>
<li><p><strong>cluster</strong> (<em>None</em>) – The result of cell clustering.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="sc_jnmf.sc_jnmf.sc_JNMF.clustering">
<code class="sig-name descname">clustering</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'hierarchical'</span></em>, <em class="sig-param"><span class="n">cluster_num</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sc_jnmf.sc_jnmf.sc_JNMF.clustering" title="Permalink to this definition">¶</a></dt>
<dd><p>This function classify the cells of input data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>method</strong> (<em>str</em><em>, </em><em>default 'hierarchical'</em>) – Select the methods for clustering. In this version, only ‘Hierarchical clustering’.
is supported.</p></li>
<li><p><strong>cluster_num</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>default None</em>) – Give the number of clusters.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sc_jnmf.sc_jnmf.sc_JNMF.factorize">
<code class="sig-name descname">factorize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">solver</span><span class="o">=</span><span class="default_value">'mu'</span></em>, <em class="sig-param"><span class="n">init</span><span class="o">=</span><span class="default_value">'random'</span></em>, <em class="sig-param"><span class="n">device</span><span class="o">=</span><span class="default_value">'gpu'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sc_jnmf.sc_jnmf.sc_JNMF.factorize" title="Permalink to this definition">¶</a></dt>
<dd><p>This function fuctorize the input gene expession matrix as ‘Joint-NMF’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>solver</strong> (<em>str</em><em>, </em><em>default 'mu'</em>) – The solver of Joint-NMF. In this version, only ‘multiplicative update’ is supported.</p></li>
<li><p><strong>init</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>defaut 'random'</em>) – The initialization of factorized matrix. In this version, only ‘random’ is supported.</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>default 'gpu'</em>) – Select the device for matrix factorization. ‘gpu’ means it is calculated using GPU,
and others means calculated using CPU.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sc_jnmf.sc_jnmf.sc_JNMF.gene_selection">
<code class="sig-name descname">gene_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rm_value1</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">rm_value2</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">threshold</span><span class="o">=</span><span class="default_value">0.06</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sc_jnmf.sc_jnmf.sc_JNMF.gene_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene filter same as SC3 clustering [Kiselev et al, 2017, nature methods(doi:10.1038/Nmeth.4236)].
This function removes gene that are either expressed (expression value &gt; rm_value1)
in less than (threshold*100)% of cells (rare genes) or expressed (expression value &gt; rm_value2)
in at least (threshold*100)% of cells (ubiquitous genes).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rm_value1</strong> (<em>int</em><em> or </em><em>float</em><em>, </em><em>default 2</em>) – threshold for counts as “gene expression” in removing rare genes.</p></li>
<li><p><strong>rm_value2</strong> (<em>int</em><em> or </em><em>float</em><em>, </em><em>default 0</em>) – threshold for counts as “gene expression” in removing ubiquitous genes.</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>default 0.06</em>) – threshold of the number of cells that satisfy the condition for removing.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="sc_jnmf.sc_jnmf.sc_JNMF.log_scale">
<code class="sig-name descname">log_scale</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sc_jnmf.sc_jnmf.sc_JNMF.log_scale" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="sc_jnmf.sc_jnmf.sc_JNMF.normalize">
<code class="sig-name descname">normalize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">norm</span><span class="o">=</span><span class="default_value">'l1'</span></em>, <em class="sig-param"><span class="n">normalize</span><span class="o">=</span><span class="default_value">'cell'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sc_jnmf.sc_jnmf.sc_JNMF.normalize" title="Permalink to this definition">¶</a></dt>
<dd><p>This function normalize the input gene expression data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>norm</strong> (<em>str</em><em>, </em><em>default 'l1'</em>) – Norm parameters for sklearn.preprocessing.Normalizer.</p></li>
<li><p><strong>normalize</strong> (<em>str</em><em>, </em><em>defaault 'cell'</em>) – Select ‘cell’ or ‘gene’ as the target of normalization.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-sc_jnmf">
<span id="module-contents"></span><h3>Module contents<a class="headerlink" href="#module-sc_jnmf" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, M.Shiga

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>